File: 205.isomorphic-strings.py
#
# @lc app=leetcode id=205 lang=python3
#
# [205] Isomorphic Strings
#

# @lc code=start
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        # Two dictionaries to store the mappings
        map_st = {}  # Mapping from s to t
        map_ts = {}  # Mapping from t to s

        for char_s, char_t in zip(s, t):
            if char_s in map_st:
                if map_st[char_s] != char_t:
                    return False
            else:
                map_st[char_s] = char_t

            if char_t in map_ts:
                if map_ts[char_t] != char_s:
                    return False
            else:
                map_ts[char_t] = char_s

        return True
        
# @lc code=end

File: 205.isomorphic-strings.java
/*
 * @lc app=leetcode id=205 lang=java
 *
 * [205] Isomorphic Strings
 */

// @lc code=start

import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean isIsomorphic(String s, String t) {
        
        if (s.length() != t.length()) {
            return false;
        }

        Map<Character, Character> s2t = new HashMap<>();
        Map<Character, Character> t2s = new HashMap<>();

        for(int i = 0; i < s.length(); i++) {

            char charOfs = s.charAt(i);
            char charOft = t.charAt(i);


            if (s2t.containsKey(charOfs)) {
                if (s2t.get(charOfs) != charOft) {
                    return false;
                }
            } else {
                s2t.put(charOfs, charOft);
            }

            if (t2s.containsKey(charOft)) {
                if (t2s.get(charOft) != charOfs) {
                    return false;
                }
            } else {
                t2s.put(charOft, charOfs);
            }
        }

        return true;
    }
}
// @lc code=end

